purity/gc/gc.go

```diff
@@ -463,72 +463,112 @@ func (c *collector) checkDirectories(dirs []*pb.Directory) {
+		return nil
+	}
+	log.Notice("Deleting %d action results, total size %s", numArs, humanize.Bytes(uint64(totalSize)))
+	ch := newProgressBar("Deleting blobs", 256)
+	defer func() {
+		close(ch)
+		time.Sleep(10 * time.Millisecond)
+	}()
+	var wg sync.WaitGroup
+	wg.Add(c.parallelism + 1)
+	step := numArs / c.parallelism
+	for i := 0; i < (c.parallelism + 1); i++ {
+		go func(ars []*ppb.Blob) {
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Hour)
+			defer cancel()
+			for _, ar := range ars {
+				log.Debug("Removing action result %s", ar.Hash)
+				if _, err := c.gcclient.Delete(ctx, &ppb.DeleteRequest{
+					ActionResults: []*ppb.Blob{ar},
+					Hard:          true,
+				}); err != nil {
+					log.Warning("Failed to delete action result %s marking as live: %w", ar.Hash, err)
+					c.liveActionResults[ar.Hash] = ar.SizeBytes
+				} else {
+					log.Debug("Deleted action result: %s", ar.Hash)
+					delete(c.actionRFs, ar.Hash)
+				}
+				ch <- 1
+			}
+			ch <- 1
+			wg.Done()
+		}(ars[step*i : min(step*(i+1), numArs)])
+	}
+	wg.Wait()
+	return nil
+}
+
+func (c *collector) RemoveBlobs() error {
+	log.Notice("Determining blobs to remove...")
+	capacity := len(c.allBlobs) - len(c.referencedBlobs)
+	blobs := make([]*ppb.Blob, 0, capacity)
+	numBlobs := 0
+	var totalSize int64
+	for hash, blob := range c.allBlobs {
 		if _, present := c.referencedBlobs[hash]; !present {
 			log.Debug("Identified blob %s for deletion", hash)
-			key := hash[:2]
-			blobs[key] = append(blobs[key], &ppb.Blob{Hash: hash, SizeBytes: size})
+			blobs = append(blobs, blob)
 			delete(c.blobRFs, hash)
-			totalSize += size
+			totalSize += blob.SizeBytes
 			numBlobs++
 		}
 	}
 	if c.dryRun {
-		log.Notice("Would delete %d action results and %d blobs, total size %s", numArs, numBlobs, humanize.Bytes(uint64(totalSize)))
+		log.Notice("Would delete %d blobs, total size %s", numBlobs, humanize.Bytes(uint64(totalSize)))
 		return nil
 	}
-	log.Notice("Deleting %d action results and %d blobs, total size %s", numArs, numBlobs, humanize.Bytes(uint64(totalSize)))
+	log.Notice("Deleting %d blobs, total size %s", numBlobs, humanize.Bytes(uint64(totalSize)))
 	ch := newProgressBar("Deleting blobs", 256)
 	defer func() {
 		close(ch)
 		time.Sleep(10 * time.Millisecond)
 	}()
-	var g multierror.Group
-	for i := 0; i < 16; i++ {
-		i := i
-		g.Go(func() error {
-			var me *multierror.Error
-			for j := 0; j < 16; j++ {
-				key := hex.EncodeToString([]byte{byte(i*16 + j)})
-				ctx, cancel := context.WithTimeout(context.Background(), 1*time.Hour)
-				defer cancel()
-				_, err := c.gcclient.Delete(ctx, &ppb.DeleteRequest{
-					Prefix:        key,
-					Blobs:         blobs[key],
-					ActionResults: ars[key],
-				})
-				if err != nil {
-					me = multierror.Append(me, err)
-				}
-				ch <- 1
+	var wg sync.WaitGroup
+	wg.Add(c.parallelism + 1)
+	step := numBlobs / c.parallelism
```

Comment:
What happens here if `numBlobs` isn't a multiple of `c.parallesism` ?

Reply:
The remainder is caught by the extra loop. For example if numBlobs = 10 and parallelism = 3. The step size would be 10 / 3 = 3 then we loop through parallelism + 1 (in this case 4) times which covers the remainder. Does that make sense?